### PR TITLE
Add `IsInteger` and  `IsFloat`; Fix `Integer` and `Float` handing with edge case

### PR DESCRIPTION
**Pull request by Emiyaaaaa**
fixed: #637 

Breaking but I think it's the corrent changes:
```ts
type Change1 = Integer<1 | 1.5>
// latest version => never // auther commented: This may be undesired behavior
// this PR        => 1

type Change2 = Integer<1e+100>
// latest version => never
// this PR        => 1e+100 // Number.IsInteger(1e+100) is true

type Change3 = Float<PositiveInfinity>
// latest version => PositiveInfinity
// this PR        => never // Infinity isn't a float.
```

And clarifies how to deal with some special number type like '1.0', '0x10' and negative number

### CODE REVIEW

File: source/numeric.d.ts [hunk 1]
```diff
@@ -1,7 +1,51 @@
+import type {Not} from './internal';
+
 export type Numeric = number | bigint;
 
 type Zero = 0 | 0n;
 
+/**
+Returns the given number if it is a float, like `1.5` or `-1.5`.
```
Comment (sindresorhus) [inline source/numeric.d.ts:8]: It doesn't return the number.

```suggestion
Returns whether the given number is a float, like `1.5` or `-1.5`.
```

File: source/numeric.d.ts [hunk 2]
```diff
@@ -1,7 +1,51 @@
...
+*/
+type IsFloat<T extends number> =
+`${T}` extends `${infer _Sign extends '' | '-'}${number}.${infer Decimal extends number}`
+	? Decimal extends Zero
+		? false
+		: true
+	: false;
+
+/**
+Returns the given number if it is an integer, like `-5`, `1` or `100`.
```
Comment (sindresorhus) [inline source/numeric.d.ts:18]:
```suggestion
Returns whether the given number is an integer, like `-5`, `1` or `100`.
```

File: source/numeric.d.ts [hunk 3]
```diff
@@ -67,7 +131,10 @@ declare function setYear<T extends number>(length: Integer<T>): void;
...
 // `${bigint}` is a type that matches a valid bigint literal without the `n` (ex. 1, 0b1, 0o1, 0x1)
 // Because T is a number and not a string we can effectively use this to filter out any numbers containing decimal points
-export type Integer<T extends number> = `${T}` extends `${bigint}` ? T : never;
+export type Integer<T extends number> =
+	T extends unknown // To distributive type
+		? IsInteger<T> extends true ? T : never
+		: never; // Never happens
 
 /**
 A `number` that is not an integer.
```
Comment (sindresorhus) [inline source/numeric.d.ts:152]: It should document the behavior for `Infinity`.

File: source/numeric.d.ts (hunk 1 above)
Comment (sindresorhus) [reply source/numeric.d.ts:8]: It should document the behavior for `Infinity`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by Emiyaaaaa** [comment]
Should we expose `IsFloat` and `IsInteger`? I think these are useful types. @sindresorhus

**Comment by sindresorhus** [comment]
Yes. I think they are useful too. Up to you whether it's done now or in a later PR.

**Pull request merged**